fdbserver/ptxn/TLogGroupVersionTracker.actor.cpp

```diff
@@ -60,9 +65,10 @@ std::map<TLogGroupID, Version> TLogGroupVersionTracker::updateGroups(const std::
 		it->second = commitVersion;
 	}
 
-	if (SERVER_KNOBS->INSERT_EMPTY_TRANSACTION || SERVER_KNOBS->BROADCAST_TLOG_GROUPS) {
+	if (SERVER_KNOBS->INSERT_EMPTY_TRANSACTION || SERVER_KNOBS->BROADCAST_TLOG_GROUPS ||
+	    returnAllGroups == UpdateAllGroups::True) {
 		for (auto& [gid, version] : versions) {
-			if (SERVER_KNOBS->BROADCAST_TLOG_GROUPS ||
+			if (returnAllGroups == UpdateAllGroups::True || SERVER_KNOBS->BROADCAST_TLOG_GROUPS ||
 			    (results.count(gid) == 0 &&
 			     commitVersion - version >= SERVER_KNOBS->LAGGING_TLOG_GROUP_VERSION_LIMIT)) {
```

Comment:
What does this `commitVersion - version` do?

Reply:
This is a feature: when a group has no commits for `LAGGING_TLOG_GROUP_VERSION_LIMIT` seconds, we insert empty commits for that group to advance its versions. In this way, different groups' version won't deviate too much, thus may potentially simply recovery by bounding group difference.

Reply:
should this condition be put in the outer `if` statement as well?

Reply:
We can't, because `version` is per group.

Reply:
Make sense. I thought this operate on a single group.